src/tools/wasm-fuzz-lattices.cpp

```diff
@@ -35,14 +42,421 @@ using namespace analysis;
 // Helps printing error messages.
 std::string LatticeComparisonNames[4] = {
   "No Relation", "Equal", "Less", "Greater"};
-std::string LatticeComparisonSymbols[4] = {"?", "=", "<", ">"};
 
 uint64_t getSeed() {
   // Return a (truly) random 64-bit value.
   std::random_device rand;
   return std::uniform_int_distribution<uint64_t>{}(rand);
 }
 
+// Actually a pointer to `L::ElementImpl`, but we erase the type to avoid
+// getting into a situation where `L` satisfying `Lattice` or `FullLattice`
+// circularly requires that `L` satisfies `Lattice` or `FullLattice`. C++ does
+// not allow concepts to depend on themselves. Also make the pointer copyable to
+// satisfy that constraint on lattice elements.
+template<typename L>
+struct RandomElement : std::unique_ptr<void, void (*)(void*)> {
+  RandomElement() = default;
+
+  RandomElement(typename L::ElementImpl&& other)
+    : std::unique_ptr<void, void (*)(void*)>(
+        new typename L::ElementImpl(std::move(other)),
+        [](void* e) { delete static_cast<typename L::ElementImpl*>(e); }) {}
+
+  RandomElement(const RandomElement& other)
+    : RandomElement([&]() {
+        auto copy = *other;
+        return copy;
+      }()) {}
```

Comment:
Can't `other` be unallocated before the lambda here is called? The lambda doesn't keep it alive AFAIK.

Reply:
This lambda is only called immediately as part of the initialization expression, so `other` will definitely still be alive. The only reason we have a lambda here is that I don't know how to make a copy of a referenced value without assigning it to some lvalue. We need a copy of `other` in the first place because we can't pass `other` directly to the other constructor because we can't move out of it.